OpenRA.Mods.Common/Traits/World/SmudgeLayer.cs

```diff
@@ -152,8 +156,14 @@ public void AddSmudge(CPos loc)
 				return;
 
 			if (hasSmoke && Game.CosmeticRandom.Next(0, 100) <= Info.SmokeChance)
+			{
+				var maxOffsetDistance = Info.MaxSmokeOffsetDistance.Length;
+				var x = Game.CosmeticRandom.Next(-maxOffsetDistance, maxOffsetDistance);
+				var y = Game.CosmeticRandom.Next(-maxOffsetDistance, maxOffsetDistance);
+				var offset = new WVec(x, y, 0);
 				world.AddFrameEndTask(w => w.Add(new SpriteEffect(
-					w.Map.CenterOfCell(loc), w, Info.SmokeImage, Info.SmokeSequences.Random(w.SharedRandom), Info.SmokePalette)));
+					w.Map.CenterOfCell(loc) + offset, w, Info.SmokeImage, Info.SmokeSequences.Random(Game.CosmeticRandom), Info.SmokePalette)));
```

Comment:
ignoring height will break this feature on TS slopes. I'd be fine with just adding a TODO comment